akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala

```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
+          val elem = grab(in)
+          try state(elem)
+          catch {
+            case e: ParsingException => fail(e.info)
+            case NotEnoughDataException =>
+              // we are missing a try/catch{continue} wrapper somewhere
+              throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+          }
+          if (output.nonEmpty) push(out, dequeue())
+          else if (!terminated) pull(in)
+          else completeStage()
+        } else completeStage()
       }
-      if (output.nonEmpty) ctx.push(dequeue())
-      else if (!terminated) ctx.pull()
-      else ctx.finish()
-    } else ctx.finish()
-
-  override def onPull(ctx: Context[Output]): SyncDirective = {
-    if (output.nonEmpty)
-      ctx.push(dequeue())
-    else if (ctx.isFinishing) {
-      if (terminated) ctx.finish()
-      else ctx.pushAndFinish(ParseError(ErrorInfo("Unexpected end of multipart entity")))
-    } else
-      ctx.pull()
-  }
-
-  override def onUpstreamFinish(ctx: Context[Output]): TerminationDirective = ctx.absorbTermination()
-
-  def tryParseInitialBoundary(input: ByteString): StateResult =
-    // we don't use boyerMoore here because we are testing for the boundary *without* a
-    // preceding CRLF and at a known location (the very beginning of the entity)
-    try {
-      if (boundary(input, 0)) {
-        val ix = boundaryLength
-        if (crlf(input, ix)) parseHeaderLines(input, ix + 2)
-        else if (doubleDash(input, ix)) terminate()
-        else parsePreamble(input, 0)
-      } else parsePreamble(input, 0)
-    } catch {
-      case NotEnoughDataException ⇒ continue(input, 0)((newInput, _) ⇒ tryParseInitialBoundary(newInput))
-    }
 
-  def parsePreamble(input: ByteString, offset: Int): StateResult =
-    try {
-      @tailrec def rec(index: Int): StateResult = {
-        val needleEnd = boyerMoore.nextIndex(input, index) + needle.length
-        if (crlf(input, needleEnd)) parseHeaderLines(input, needleEnd + 2)
-        else if (doubleDash(input, needleEnd)) terminate()
-        else rec(needleEnd)
+      override def onPull(): Unit = {
+        if (output.nonEmpty)
+          push(out, dequeue())
+        else if (isClosed(in)) {
+          if (!terminated) push(out, ParseError(ErrorInfo("Unexpected end of multipart entity")))
+          completeStage()
+        } else pull(in)
       }
-      rec(offset)
-    } catch {
-      case NotEnoughDataException ⇒ continue(input.takeRight(needle.length + 2), 0)(parsePreamble)
-    }
 
-  @tailrec def parseHeaderLines(input: ByteString, lineStart: Int, headers: ListBuffer[HttpHeader] = ListBuffer[HttpHeader](),
-                                headerCount: Int = 0, cth: Option[`Content-Type`] = None): StateResult = {
-    def contentType =
-      cth match {
-        case Some(x) ⇒ x.contentType
-        case None    ⇒ defaultContentType
+      override def onUpstreamFinish(): Unit = {
+        if (isAvailable(out)) onPull()
```

Comment:
Different behavior from the original stage which did just absorb termination? We are not sure there has been a pull from downstream as upstream finishes, so if output is `nonEmpty` this would lead to trying to push to non-ready downstream.